docs/specs/user-export.md

```diff
@@ -0,0 +1,284 @@
+# User Export
+
+User Export allows the developer to bulk export all users from Authgear to a file.
+
+## About User Export
+
+- It is not a synchronous operation. The export is created and runs in the background. The developer can query the status of it.
+
+## Create an export
+
+The endpoint is `POST /_api/admin/users/export`.
+
+- The endpoint requires Admin API JWT token to access.
+- This endpoint is added to Admin API server.
+- It is not part of the GraphQL API.
+- At most 1 running export at any given moment for a project.
+- The rate limit of creating a export is 24 per 24h. This means failed requests do not count.
+- A pending export lasts for 24h before it expires.
+
+### The request body of Create an export
+
+```
+{
+  "format": "ndjson",
+  "csv": {
+    "fields": [
+      {
+        "json_pointer": "/sub",
+        "field_name": "user_id"
+      }
+    ]
+  }
+}
+```
+
+- `format`: Required. It must be `ndjson` or `csv`.
+  - `ndjson`: The output is a ndjson file. See https://github.com/ndjson/ndjson-spec
+  - `csv`: The output is a CSV file. See https://datatracker.ietf.org/doc/html/rfc4180
+- `csv.fields`: Required when `format` is `csv`. It must be an non-empty array.
+  - `csv.fields.json_pointer`: Required. Select which field in the record to output. It must be a JSON pointer. See https://datatracker.ietf.org/doc/html/rfc6901 and [The record format](#the-record-format)
+  - `csv.fields.field_name`: Optional. If it is not given, then the field name is derived from `json_pointer` by taking the last reference token. For example, if the JSON pointers are `/sub` and `/address/formatted`, then the field names are `sub` and `formatted` respectively. If the field names are not unique, it is an error when the export is initiated. An error is immediately returned in this case, the import job is not created. See [The error response of Create an export](#the-error-response-of-create-an-export)
```

Comment:
In the example given, `/address/formatted` will automatically convert to `formatted` sounds quite uncomfortable, that meant `field_name` are always required.

Alternatively:

1. Instead of taking the last reference token, use a conversion of removing the first `/` and converting`/` to `-` instead? E.g. `/address/formatted` to `address-formatted`
2. For all default fields, give a reasonable default; And only use last reference token for fields without default.

The goal should be `csv.fields.field_name` is really "Optional".